ray-curator/ray_curator/backends/experimental/ray_data/executor.py

```diff
@@ -0,0 +1,111 @@
+        logger.warning("Ray Data executor is experimental and might not work as expected.")
+
+    def execute(self, stages: list["ProcessingStage"], initial_tasks: list[Task] | None = None) -> list[Task]:
+        """Execute the pipeline stages using Ray Data.
+
+        Args:
+            stages (list[ProcessingStage]): List of processing stages to execute
+            initial_tasks (list[Task], optional): Initial tasks to process (can be None for empty start)
+
+        Returns:
+            list[Task]: List of final processed tasks
+        """
+        if not stages:
+            return []
+
+        logger.info(f"Starting Ray Data pipeline with {len(stages)} stages")
+
+        # Initialize with initial tasks if provided, otherwise start with EmptyTask
+        tasks: list[Task] = initial_tasks if initial_tasks else [EmptyTask]
+
+        # Convert tasks to dataset
+        current_dataset = self._tasks_to_dataset(tasks)
+        logger.info(f"Initial dataset size: {current_dataset.count()}")
+
+        try:
+            # Process through each stage
+            for i, stage in enumerate(stages):
+                # TODO: add pipeline level config for verbosity
+                logger.info(f"Processing stage {i + 1}/{len(stages)}: {stage}")
+                logger.info(f"  CPU cores: {stage.resources.cpus}, GPU ratio: {stage.resources.gpus}")
+
+                # Create adapter for this stage
+                adapter = RayDataStageAdapter(stage)
+
+                # Apply stage transformation
+                current_dataset = adapter.process_dataset(current_dataset)
+
+                # Log progress
+                logger.info(f"  Stage {i + 1} completed")
+
+        except Exception as e:
+            logger.error(f"Error during pipeline execution: {e}")
+            raise
+
+        else:
+            # Convert final dataset back to tasks
+            # TODO: add pipeline configuration to check if user wants to return last stages output to driver
+            logger.info(f"Converting final dataset to tasks -> {current_dataset=}")
+            final_tasks = self._dataset_to_tasks(current_dataset)
+            logger.info(f"Pipeline completed. Final results: {len(final_tasks)} tasks")
+
+            return final_tasks
+
+    def _tasks_to_dataset(self, tasks: list[Task]) -> Dataset:
+        """Convert list of tasks to Ray Data dataset.
+
+        Args:
+            tasks: List of Task objects
+
+        Returns:
+            Ray Data dataset containing Task objects directly
+        """
+        # Create Ray Data dataset directly from Task objects
+        return ray.data.from_items(tasks)
+
+    def _dataset_to_tasks(self, dataset: Dataset) -> list[Task]:
+        """Convert Ray Data dataset back to list of tasks.
+
+        Args:
+            dataset: Ray Data dataset containing Task objects
+
+        Returns:
+            List of Task objects
+        """
+        # Get all items from dataset
+        items = dataset.take_all()
```

Comment:
Similar caveat. Take all will first materialize this whole dataset in object store memory and then possibly bring it to the driver process (similar to `compute` in dask)